packages/adminrouter/extra/src/lib/service.lua

```diff
@@ -107,6 +107,43 @@ local function resolve_via_marathon_apps_state(service_name, marathon_cache)
     return true, nil, nil
 end
 
+local function requrl_rewriting_required(service_name, marathon_cache)
+    -- Determine if the upstream request should contain the
+    -- `/service/<service_name>` part of the URL path or not.
+    --
+    -- Arguments:
+    --   service_name (string): service name that should be resolved
+    --
+    -- Returns:
+    --  A list with:
+    --  - true/false depending on whether requrl rewriting is enabled or not
+    --  - err_code, err_text - if an error occured these will be HTTP status
+    --    and error text that should be sent to the client. `nil` otherwise
+    --
+    if service_name == 'marathon' or service_name == 'metronome' then
+      -- These two services should always have the request ulr rewriting
+      -- enabled, no matter the marathon_cache, as they are resovled via Mesos
+      -- anyway.
+      return true, nil, nil
+    end
+
+    if marathon_cache == nil then
+        -- In theory this should never be reached for non-marathon and
+        -- non-metronome services, as
+        -- `resolve_via_marathon_apps_state` is always called first and once
+        -- the cache value has been set, it is never removed. Still - just in
+        -- case.
+        return nil, ngx.HTTP_SERVICE_UNAVAILABLE, "503 Service Unavailable: invalid Marathon svcapps cache while resolving requrl rewriting"
```

Comment:
This line causes Nginx to return 503 to the client with the given message. The `return` statement is just a convention/idiom in the openresty world, it is meant to signify that the request processing ends here. The calling function's code is never resumed.

Reply:
I was not clear.

This is a line that we the comment says we should not hit - right? Yet we have it anyways. I'm not 100% sure why - I'm very wary of adding code that we can't hit in a test. That said, if we are to have this code, and we get a report that someone saw this error, what do we do?

Reply:
This error will be caught earlier by [1]. I have put the same error checking here as well because:
* if in the future the order in which the code is executed changes, lack of error checking here will become a serious bug
* if for some reason, the code starts resetting/invalidating stale cache then we will start having a nasty race condition here which will be resulting in a Lua stacktraces from time to time.
I agree that we should not commit dead code into dc/os, but it seems like a lesser evil.

As for your second question - we investigate why the cache was unavailable. The place when this error is generated does not matter much and the same error message is IIRC emitted in few other places.

[1] https://github.com/dcos/dcos/blob/16ced28a63284de75096e17fa953af08905cd91d/packages/adminrouter/extra/src/lib/service.lua#L96-L98

Reply:
Not at the moment, the error condition that we check for here will always be caught in [1] first.

[1] https://github.com/dcos/dcos/blob/16ced28a63284de75096e17fa953af08905cd91d/packages/adminrouter/extra/src/lib/service.lua#L96-L98

Reply:
@jgehrcke tl;dr - We won't hit this code at the time of writing because another function which is called first does a check that the required condition is not true. There is no current infrastructure for unit testing this function. Changing another function, or the order in which functions are called, may cause us to hit a condition for which raising an error here will avoid a bug.

What do you think about keeping this code in?

Reply:
I think this PR needs to improve its error handling story/concept.

If we give up and say "we have no idea how our code behaves" (I am not really in favor of that) and want to prepare ourselves for that case then I don't think it is wise to to _not_ log all relevant detail (as far as I read the current code it would simply emit a 503 response, but there would not be a corresponding informative log message).

That is, _iff_ we want to prepare for this rather fatal scenario, then I'd like to see a precise 'error' level log message, stack traces, etc in the Admin Router log.

> which will be resulting in a Lua stacktraces from time to time.

If one handles errors, one needs to handle them very consciously. I think letting things blow up would be better than the current code: it would log a very useful lua stack trace :-) and it would return a '500 internal server error' to the user agent which is about equally useful as the current HTTP error response.

Not adding these 'paranoid' code paths seems like the better thing to do, _unless_ we have better error handling in place than the default error handling (again, which is log a lua stack trace, and return a '500 internal server error').